wgengine/magicsock: TestNoDiscoKey test timeout

### What is the issue?

CI failure in https://github.com/tailscale/tailscale/pull/3735.

cc @bradfitz 

```
panic: test timed out after 10m0s

goroutine 846 [running]:
testing.(*M).startAlarm.func1()
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/testing/testing.go:1788 +0x8e
created by time.goFunc
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/time/sleep.go:180 +0x31

goroutine 1 [chan receive, 9 minutes]:
testing.(*T).Run(0xc0000896c0, {0x188638e, 0x132ddf3}, 0x18ae980)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/testing/testing.go:1307 +0x375
testing.runTests.func1(0xc0000896c0)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/testing/testing.go:1598 +0x6e
testing.tRunner(0xc0000896c0, 0xc0000c3d18)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/testing/testing.go:1259 +0x102
testing.runTests(0xc0000d2180, {0x179b660, 0x13, 0x13}, {0x135846d, 0x1886746, 0x1c1cc00})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/testing/testing.go:1596 +0x43f
testing.(*M).Run(0xc0000d2180)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/testing/testing.go:1504 +0x51d
main.main()
	_testmain.go:83 +0x14b

goroutine 359 [chan send, 9 minutes]:
tailscale.com/wgengine/magicsock.(*connBind).Close(0xc0000047c8)
	D:/a/tailscale/tailscale/wgengine/magicsock/magicsock.go:2650 +0x11e
golang.zx2c4.com/wireguard/device.closeBindLocked(0xc00020ec80)
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/device.go:416 +0x39
golang.zx2c4.com/wireguard/device.(*Device).BindClose(0xc00020ec80)
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/device.go:522 +0x36
golang.zx2c4.com/wireguard/device.(*Device).downLocked(0xc00020ec80)
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/device.go:196 +0x33
golang.zx2c4.com/wireguard/device.(*Device).Close(0xc00020ec80)
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/device.go:366 +0xf8
tailscale.com/wgengine/magicsock.(*magicStack).Close(0xc000072550)
	D:/a/tailscale/tailscale/wgengine/magicsock/magicsock_test.go:205 +0x25
tailscale.com/wgengine/magicsock.TestNoDiscoKey(0xc00045c000)
	D:/a/tailscale/tailscale/wgengine/magicsock/magicsock_test.go:652 +0x551
testing.tRunner(0xc00045c000, 0x18ae980)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/testing/testing.go:1306 +0x35a

goroutine 369 [semacquire, 9 minutes]:
sync.runtime_Semacquire(0x175d950)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/runtime/sema.go:56 +0x25
sync.(*WaitGroup).Wait(0xc000184280)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/sync/waitgroup.go:130 +0x71
golang.zx2c4.com/wireguard/device.newInboundQueue.func1()
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/channels.go:50 +0x28
created by golang.zx2c4.com/wireguard/device.newInboundQueue
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/channels.go:49 +0xb2

goroutine 524 [select, 9 minutes]:
tailscale.com/wgengine/magicsock.(*Conn).runDerpWriter(0xc000472b00, {0x1944038, 0xc000059e00}, 0xc00040a118, 0xc000320ba0, 0xc00009e0c0, 0xc0000839e0)
	D:/a/tailscale/tailscale/wgengine/magicsock/magicsock.go:1630 +0x1d4
created by tailscale.com/wgengine/magicsock.(*Conn).derpWriteChanOfAddr
	D:/a/tailscale/tailscale/wgengine/magicsock/magicsock.go:1421 +0xfd3

goroutine 372 [chan receive, 9 minutes]:
golang.zx2c4.com/wireguard/device.(*Device).RoutineHandshake(0xc00020ec80, 0x1)
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/receive.go:244 +0x19b
created by golang.zx2c4.com/wireguard/device.NewDevice
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/device.go:316 +0x2f8

goroutine 361 [IO wait]:
internal/poll.runtime_pollWait(0x18cd16e3148, 0x72)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/runtime/netpoll.go:234 +0x89
internal/poll.(*pollDesc).wait(0xc00020e398, 0xc00020e3a8, 0x0)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/internal/poll/fd_poll_runtime.go:84 +0x32
internal/poll.execIO(0xc00020e298, 0x18adf60)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/internal/poll/fd_windows.go:175 +0xe5
internal/poll.(*FD).ReadFrom(0xc00020e280, {0xc00005a000, 0x12d4ee5, 0x2030000})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/internal/poll/fd_windows.go:581 +0x155
net.(*netFD).readFrom(0xc00020e280, {0xc00005a000, 0x0, 0xc000491e48})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/fd_posix.go:62 +0x29
net.(*UDPConn).readFrom(0x1c1d1a0, {0xc00005a000, 0x0, 0xc000022940}, 0xc0003233b0)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/udpsock_posix.go:47 +0x3e
net.(*UDPConn).readFromUDP(0xc0000c60a0, {0xc00005a000, 0x20, 0xc000491ebe}, 0x4)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/udpsock.go:116 +0x31
net.(*UDPConn).ReadFrom(0xc0000c60a0, {0xc00005a000, 0x10000, 0x10000})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/udpsock.go:125 +0x50
tailscale.com/net/stun/stuntest.runSTUN({0x19527e0, 0xc00045c000}, {0x194a760, 0xc0000c60a0}, 0xc0000a4540, 0xc0004337c0)
	D:/a/tailscale/tailscale/net/stun/stuntest/stuntest.go:60 +0xc6
created by tailscale.com/net/stun/stuntest.ServeWithPacketListener
	D:/a/tailscale/tailscale/net/stun/stuntest/stuntest.go:48 +0x25d

goroutine 380 [semacquire, 9 minutes]:
sync.runtime_SemacquireMutex(0x12d8b2d, 0x78, 0x0)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/runtime/sema.go:71 +0x25
sync.(*Mutex).lockSlow(0xc0000047d0)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/sync/mutex.go:138 +0x165
sync.(*Mutex).Lock(...)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/sync/mutex.go:81
tailscale.com/wgengine/magicsock.(*connBind).Closed(0x4)
	D:/a/tailscale/tailscale/wgengine/magicsock/magicsock.go:2656 +0x57
tailscale.com/wgengine/magicsock.(*connBind).receiveDERP(0xc0000047c8, {0xc000250000, 0x7e0, 0x7e0})
	D:/a/tailscale/tailscale/wgengine/magicsock/magicsock.go:1722 +0x1ae
golang.zx2c4.com/wireguard/device.(*Device).RoutineReceiveIncoming(0xc00020ec80, 0xc00008bad0)
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/receive.go:94 +0x18f
created by golang.zx2c4.com/wireguard/device.(*Device).BindUpdate
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/device.go:513 +0x439

goroutine 368 [semacquire, 9 minutes]:
sync.runtime_Semacquire(0x175d948)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/runtime/sema.go:56 +0x25
sync.(*WaitGroup).Wait(0xc000184280)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/sync/waitgroup.go:130 +0x71
golang.zx2c4.com/wireguard/device.newOutboundQueue.func1()
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/channels.go:32 +0x28
created by golang.zx2c4.com/wireguard/device.newOutboundQueue
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/channels.go:31 +0xb2

goroutine 360 [IO wait, 9 minutes]:
internal/poll.runtime_pollWait(0x18cd16e3970, 0x72)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/runtime/netpoll.go:234 +0x89
internal/poll.(*pollDesc).wait(0xc00004cae0, 0x12ce214, 0x0)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/internal/poll/fd_poll_runtime.go:84 +0x32
internal/poll.execIO(0xc00020e018, 0xc000387bb0)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/internal/poll/fd_windows.go:175 +0xe5
internal/poll.(*FD).acceptOne(0xc00020e000, 0x304, {0xc0003583c0, 0x40, 0x1c79ed8}, 0xc000387be0)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/internal/poll/fd_windows.go:810 +0x6d
internal/poll.(*FD).Accept(0xc00020e000, 0xc000387d88)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/internal/poll/fd_windows.go:844 +0x1d6
net.(*netFD).accept(0xc00020e000)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/fd_windows.go:139 +0x65
net.(*TCPListener).accept(0xc000004648)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/tcpsock_posix.go:140 +0x28
net.(*TCPListener).Accept(0xc000004648)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/tcpsock.go:262 +0x3d
crypto/tls.(*listener).Accept(0xc000004798)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/crypto/tls/tls.go:66 +0x2d
net/http.(*Server).Serve(0xc0002280e0, {0x1940bf0, 0xc000004798})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/http/server.go:3002 +0x394
net/http/httptest.(*Server).goServe.func1()
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/http/httptest/server.go:308 +0x6a
created by net/http/httptest.(*Server).goServe
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/http/httptest/server.go:306 +0x6f

goroutine 374 [chan receive, 9 minutes]:
golang.zx2c4.com/wireguard/device.(*Device).RoutineDecryption(0xc00020ec80, 0x0)
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/receive.go:212 +0x1a8
created by golang.zx2c4.com/wireguard/device.NewDevice
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/device.go:315 +0x3b0

goroutine 370 [chan receive, 9 minutes]:
golang.zx2c4.com/wireguard/device.(*Device).RoutineEncryption(0xc00020ec80, 0xc00009ea68)
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/send.go:376 +0x1c5
created by golang.zx2c4.com/wireguard/device.NewDevice
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/device.go:314 +0x365

goroutine 366 [select, 9 minutes]:
golang.zx2c4.com/wireguard/ratelimiter.(*Ratelimiter).Init.func1()
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/ratelimiter/ratelimiter.go:70 +0x94
created by golang.zx2c4.com/wireguard/ratelimiter.(*Ratelimiter).Init
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/ratelimiter/ratelimiter.go:66 +0x17d

goroutine 501 [semacquire, 9 minutes]:
sync.runtime_SemacquireMutex(0x94, 0x74, 0xc000586a74)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/runtime/sema.go:71 +0x25
sync.(*RWMutex).RLock(...)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/sync/rwmutex.go:63
golang.zx2c4.com/wireguard/device.(*Peer).SendBuffer(0xc0000d5180, {0xc000586a00, 0x94, 0x94})
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/peer.go:120 +0x72
golang.zx2c4.com/wireguard/device.(*Peer).SendHandshakeInitiation(0xc0000d5180, 0xd6)
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/send.go:129 +0x3d1
golang.zx2c4.com/wireguard/device.expiredNewHandshake(0xc0000d5180)
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/timers.go:132 +0x10a
golang.zx2c4.com/wireguard/device.(*Peer).NewTimer.func1()
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/timers.go:43 +0xcd
created by time.goFunc
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/time/sleep.go:180 +0x31

goroutine 373 [chan receive, 9 minutes]:
golang.zx2c4.com/wireguard/device.(*Device).RoutineEncryption(0xc00020ec80, 0xc000082780)
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/send.go:376 +0x1c5
created by golang.zx2c4.com/wireguard/device.NewDevice
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/device.go:314 +0x365

goroutine 371 [chan receive, 9 minutes]:
golang.zx2c4.com/wireguard/device.(*Device).RoutineDecryption(0xc00020ec80, 0x1)
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/receive.go:212 +0x1a8
created by golang.zx2c4.com/wireguard/device.NewDevice
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/device.go:315 +0x3b0

goroutine 375 [chan receive, 9 minutes]:
golang.zx2c4.com/wireguard/device.(*Device).RoutineHandshake(0xc00020ec80, 0x2)
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/receive.go:244 +0x19b
created by golang.zx2c4.com/wireguard/device.NewDevice
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/device.go:316 +0x2f8

goroutine 523 [IO wait]:
internal/poll.runtime_pollWait(0x18cd16e3888, 0x72)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/runtime/netpoll.go:234 +0x89
internal/poll.(*pollDesc).wait(0xc000182f08, 0x0, 0x0)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/internal/poll/fd_poll_runtime.go:84 +0x32
internal/poll.execIO(0xc000267b98, 0x18adf58)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/internal/poll/fd_windows.go:175 +0xe5
internal/poll.(*FD).Read(0xc000267b80, {0xc0003d2a00, 0x6e0, 0x6e0})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/internal/poll/fd_windows.go:441 +0x25f
net.(*netFD).Read(0xc000267b80, {0xc0003d2a00, 0xc0003d2a05, 0x16})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/fd_posix.go:56 +0x29
net.(*conn).Read(0xc0000062f8, {0xc0003d2a00, 0x6, 0xc0001830d8})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/net.go:183 +0x45
crypto/tls.(*atLeastReader).Read(0xc0000043a8, {0xc0003d2a00, 0xc00045cd00, 0x12cadcd})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/crypto/tls/conn.go:777 +0x3d
bytes.(*Buffer).ReadFrom(0xc0004185f8, {0x1935ca0, 0xc0000043a8})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/bytes/buffer.go:204 +0x98
crypto/tls.(*Conn).readFromUntil(0xc000418380, {0x18cd1702e18, 0xc0000062f8}, 0x1657f7b)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/crypto/tls/conn.go:799 +0xe5
crypto/tls.(*Conn).readRecordOrCCS(0xc000418380, 0x0)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/crypto/tls/conn.go:606 +0x112
crypto/tls.(*Conn).readRecord(...)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/crypto/tls/conn.go:574
crypto/tls.(*Conn).Read(0xc000418380, {0xc0003e8000, 0x1000, 0x14c078e})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/crypto/tls/conn.go:1277 +0x16f
bufio.(*Reader).fill(0xc0003e64e0)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/bufio/bufio.go:101 +0x103
bufio.(*Reader).ReadByte(0xc0003e64e0)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/bufio/bufio.go:253 +0x2c
tailscale.com/derp.readFrameHeader(0xc0704da2298125e4)
	D:/a/tailscale/tailscale/derp/derp.go:161 +0x1e
tailscale.com/derp.(*Client).recvTimeout(0xc0002fa270, 0x1944070)
	D:/a/tailscale/tailscale/derp/derp_client.go:469 +0x218
tailscale.com/derp.(*Client).Recv(...)
	D:/a/tailscale/tailscale/derp/derp_client.go:442
tailscale.com/derp/derphttp.(*Client).RecvDetail(0xc0704d84298125e4)
	D:/a/tailscale/tailscale/derp/derphttp/derphttp_client.go:900 +0x7c
tailscale.com/wgengine/magicsock.(*Conn).runDerpReader(0xc000472b00, {0x1944038, 0xc000059e00}, {{{0x0, 0xc000472b00}, 0xc00009e0c0}, 0x6740}, 0xc00040c540, 0x0, 0xc0000839e0)
	D:/a/tailscale/tailscale/wgengine/magicsock/magicsock.go:1514 +0x513
created by tailscale.com/wgengine/magicsock.(*Conn).derpWriteChanOfAddr
	D:/a/tailscale/tailscale/wgengine/magicsock/magicsock.go:1420 +0xee9

goroutine 549 [select, 3 minutes]:
tailscale.com/derp.(*sclient).sendLoop(0xc000197900, {0x1944038, 0xc0003ea480})
	D:/a/tailscale/tailscale/derp/derp_server.go:1393 +0x609
tailscale.com/derp.(*sclient).run.func1()
	D:/a/tailscale/tailscale/derp/derp_server.go:702 +0x25
golang.org/x/sync/errgroup.(*Group).Go.func1()
	C:/Users/runneradmin/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x67
created by golang.org/x/sync/errgroup.(*Group).Go
	C:/Users/runneradmin/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:54 +0x92

goroutine 367 [semacquire, 9 minutes]:
sync.runtime_Semacquire(0x0)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/runtime/sema.go:56 +0x25
sync.(*WaitGroup).Wait(0x171d5e6)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/sync/waitgroup.go:130 +0x71
golang.zx2c4.com/wireguard/device.newHandshakeQueue.func1()
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/channels.go:68 +0x28
created by golang.zx2c4.com/wireguard/device.newHandshakeQueue
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/channels.go:67 +0xb2

goroutine 546 [IO wait, 9 minutes]:
internal/poll.runtime_pollWait(0x18cd16e35d0, 0x72)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/runtime/netpoll.go:234 +0x89
internal/poll.(*pollDesc).wait(0x7, 0x5, 0x0)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/internal/poll/fd_poll_runtime.go:84 +0x32
internal/poll.execIO(0xc00039c018, 0x18adf58)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/internal/poll/fd_windows.go:175 +0xe5
internal/poll.(*FD).Read(0xc00039c000, {0xc000545200, 0x205, 0x205})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/internal/poll/fd_windows.go:441 +0x25f
net.(*netFD).Read(0xc00039c000, {0xc000545200, 0xc000545205, 0x17})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/fd_posix.go:56 +0x29
net.(*conn).Read(0xc0000062f0, {0xc000545200, 0x132cac0, 0xc00017f128})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/net.go:183 +0x45
crypto/tls.(*atLeastReader).Read(0xc00009f6c8, {0xc000545200, 0x0, 0x12cadcd})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/crypto/tls/conn.go:777 +0x3d
bytes.(*Buffer).ReadFrom(0xc000418278, {0x1935ca0, 0xc00009f6c8})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/bytes/buffer.go:204 +0x98
crypto/tls.(*Conn).readFromUntil(0xc000418000, {0x18cd1702e18, 0xc0000062f0}, 0x205)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/crypto/tls/conn.go:799 +0xe5
crypto/tls.(*Conn).readRecordOrCCS(0xc000418000, 0x0)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/crypto/tls/conn.go:606 +0x112
crypto/tls.(*Conn).readRecord(...)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/crypto/tls/conn.go:574
crypto/tls.(*Conn).Read(0xc000418000, {0xc0003fa000, 0x1000, 0xc0704d04a9e6efb4})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/crypto/tls/conn.go:1277 +0x16f
net/http.(*connReader).Read(0xc0003234d0, {0xc0003fa000, 0x1000, 0x1000})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/http/server.go:780 +0x16d
bufio.(*Reader).fill(0xc0003e6780)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/bufio/bufio.go:101 +0x103
bufio.(*Reader).ReadByte(0xc0003e6780)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/bufio/bufio.go:253 +0x2c
tailscale.com/derp.readFrameHeader(0xc000197900)
	D:/a/tailscale/tailscale/derp/derp.go:161 +0x1e
tailscale.com/derp.(*sclient).run(0xc000197900, {0x1944038, 0xc0003ea400})
	D:/a/tailscale/tailscale/derp/derp_server.go:711 +0x16f
tailscale.com/derp.(*Server).accept(0xc0003c6000, {0x18cd1709160, 0xc000418000}, 0xc000037ff0, {0xc000417190, 0x188146b}, 0x3)
	D:/a/tailscale/tailscale/derp/derp_server.go:687 +0xa9e
tailscale.com/derp.(*Server).Accept(0xc0003c6000, {0x18cd1709160, 0xc000418000}, 0x6f, {0xc000417190, 0xf})
	D:/a/tailscale/tailscale/derp/derp_server.go:419 +0x1f5
tailscale.com/derp/derphttp.Handler.func1({0x1940fb0, 0xc00040c700}, 0xc00046f500)
	D:/a/tailscale/tailscale/derp/derphttp/derphttp_server.go:59 +0x333
net/http.HandlerFunc.ServeHTTP(0x0, {0x1940fb0, 0xc00040c700}, 0x1328ef5)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/http/server.go:2047 +0x2f
net/http.serverHandler.ServeHTTP({0xc0003234d0}, {0x1940fb0, 0xc00040c700}, 0xc00046f500)
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/http/server.go:2879 +0x43b
net/http.(*conn).serve(0xc00020d720, {0x19440e0, 0xc0000ba960})
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/http/server.go:1930 +0xb08
created by net/http.(*Server).Serve
	C:/hostedtoolcache/windows/go/1.17.6/x64/src/net/http/server.go:3034 +0x4e8

goroutine 447 [chan receive, 9 minutes]:
golang.zx2c4.com/wireguard/device.(*Peer).RoutineSequentialSender(0xc0000d5180)
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/send.go:418 +0x11a
created by golang.zx2c4.com/wireguard/device.(*Peer).Start
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/peer.go:198 +0x253

goroutine 448 [chan receive, 9 minutes]:
golang.zx2c4.com/wireguard/device.(*Peer).RoutineSequentialReceiver(0xc0000d5180)
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/receive.go:406 +0x145
created by golang.zx2c4.com/wireguard/device.(*Peer).Start
	C:/Users/runneradmin/go/pkg/mod/golang.zx2c4.com/wireguard@v0.0.0-20211116201604-de7c702ace45/device/peer.go:199 +0x295
exit status 2
FAIL	tailscale.com/wgengine/magicsock	600.057s
```



### Steps to reproduce

_No response_

### Are there any recent changes that introduced the issue?

_No response_

### OS

_No response_

### OS version

_No response_

### Tailscale version

_No response_

### Bug report

_No response_

--- a/wgengine/magicsock/magicsock.go
+++ b/wgengine/magicsock/magicsock.go
@@ -265,6 +265,7 @@ type Conn struct {
 	stunReceiveFunc atomic.Value // of func(p []byte, fromAddr *net.UDPAddr)
 
 	// derpRecvCh is used by receiveDERP to read DERP messages.
+	// It must have buffer size > 0; see issue 3736.
 	derpRecvCh chan derpReadResult
 
 	// bind is the wireguard-go conn.Bind for Conn.
@@ -529,7 +530,7 @@ func (o *Options) derpActiveFunc() func() {
 // of NewConn. Mostly for tests.
 func newConn() *Conn {
 	c := &Conn{
-		derpRecvCh:   make(chan derpReadResult),
+		derpRecvCh:   make(chan derpReadResult, 1), // must be buffered, see issue 3736
 		derpStarted:  make(chan struct{}),
 		peerLastDerp: make(map[key.NodePublic]int),
 		peerMap:      newPeerMap(),
@@ -2647,6 +2648,7 @@ func (c *connBind) Close() error {
 	}
 	// Send an empty read result to unblock receiveDERP,
 	// which will then check connBind.Closed.
+	// connBind.Closed takes c.mu, but c.derpRecvCh is buffered.
 	c.derpRecvCh <- derpReadResult{}
 	return nil
 }

